analyzer/emerald/emerald.go

```diff
@@ -0,0 +1,272 @@
+package emerald
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/iancoleman/strcase"
+	"github.com/jackc/pgx/v4"
+	oasisConfig "github.com/oasisprotocol/oasis-sdk/client-sdk/go/config"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/oasisprotocol/oasis-indexer/analyzer"
+	"github.com/oasisprotocol/oasis-indexer/analyzer/modules"
+	"github.com/oasisprotocol/oasis-indexer/analyzer/util"
+	"github.com/oasisprotocol/oasis-indexer/config"
+	"github.com/oasisprotocol/oasis-indexer/log"
+	"github.com/oasisprotocol/oasis-indexer/metrics"
+	"github.com/oasisprotocol/oasis-indexer/storage"
+	"github.com/oasisprotocol/oasis-indexer/storage/oasis"
+)
+
+const (
+	emerald = analyzer.RuntimeEmerald
+
+	emeraldMainDamaskName = "emerald_main_damask"
+)
+
+// Main is the main Analyzer for the Emerald Runtime.
+type Main struct {
+	cfg     analyzer.RuntimeConfig
+	qf      analyzer.QueryFactory
+	target  storage.TargetStorage
+	logger  *log.Logger
+	metrics metrics.DatabaseMetrics
+
+	moduleHandlers []modules.ModuleHandler
+}
+
+// NewMain returns a new main analyzer for the Emerald Runtime.
+func NewMain(cfg *config.AnalyzerConfig, target storage.TargetStorage, logger *log.Logger) (*Main, error) {
+	ctx := context.Background()
+
+	// Initialize source storage.
+	networkCfg := oasisConfig.Network{
+		ChainContext: cfg.ChainContext,
+		RPC:          cfg.RPC,
+	}
+	factory, err := oasis.NewClientFactory(ctx, &networkCfg)
+	if err != nil {
+		logger.Error("error creating client factory",
+			"err", err,
+		)
+		return nil, err
+	}
+
+	network, err := analyzer.FromChainContext(cfg.ChainContext)
+	if err != nil {
+		return nil, err
+	}
+
+	id, err := emerald.ID(network)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := factory.Runtime(id)
+	if err != nil {
+		logger.Error("error creating runtime client",
+			"err", err,
+		)
+		return nil, err
+	}
+	roundRange := analyzer.RoundRange{
+		From: uint64(cfg.From),
+		To:   uint64(cfg.To),
+	}
+	ac := analyzer.RuntimeConfig{
+		Range:  roundRange,
+		Source: client,
+	}
+
+	qf := analyzer.NewQueryFactory(strcase.ToSnake(cfg.ChainID), emerald.String())
+
+	return &Main{
+		cfg:     ac,
+		qf:      qf,
+		target:  target,
+		logger:  logger,
+		metrics: metrics.NewDefaultDatabaseMetrics(emeraldMainDamaskName),
+
+		// module handlers
```

Comment:
They can't be used for consensus, since consensus is its own beast and its storage backend architecture differs from ParaTimes and their modules architecture.

But these can are indeed intended for reuse across other ParaTimes.